File: benchmark_scripts/nbody-mpi/jl/nbody.jl
using Random
using Base.Threads

using MPI
MPI.Init()

comm = MPI.COMM_WORLD

DELTA_TIME=1.0
CONST_G=1.0

mutable struct Particle
    mass::Float64
    posx::Float64
    posy::Float64
    velx::Float64
    vely::Float64
    accx::Float64
    accy::Float64
end

function distance(p1::Particle, p2::Particle)::Float64
    return sqrt((p1.posx - p2.posx)^2 + (p1.posy - p2.posy)^2)
end

function update_position!(p::Particle)
    p.posx = p.posx + p.velx * DELTA_TIME;
    p.posy = p.posy + p.vely * DELTA_TIME; 
end

function update_velocity!(p::Particle)
    p.velx = p.velx + p.accx/p.mass * DELTA_TIME;
    p.vely = p.vely + p.accy/p.mass * DELTA_TIME; 
end

function update_acceleration!(p1::Particle, p2::Particle)
    dist = distance(p1, p2);
    f = CONST_G*(p1.mass * p2.mass)/(dist^2);

    p1.accx = p1.accx + f * (p2.posx - p1.posx) / dist 
    p1.accy = p1.accy + f * (p2.posy - p1.posy) / dist
end

function reset_force!(p::Particle)
    p.accx = 0.0
    p.accy = 0.0
end

function main()
    nb_particles = parse(Int64, ARGS[1])
    nb_timesteps = parse(Int64, ARGS[2])
    before = time()
    Random.seed!(0)
    rank = MPI.Comm_rank(comm)
    world = MPI.Comm_size(comm)
    particles = nothing
    if rank == 0
        particles = 1:nb_particles .|> (x -> Particle(rand(Float64, 1) * 100.0..., rand(Float64, 6)...))
    end

    sizes = fill(0, world)
    offsets = fill(0, world)
    cum_sum = 0
    for k in 1:world
        sizes[k] = div(nb_particles, world)
        offsets[k] = cum_sum
        cum_sum += sizes[k]
    end
    sizes[world] = nb_particles - (world-1) * (div(nb_particles, world))
    particles = MPI.bcast(particles, 0, comm)
    my_particles = Vector{Particle}(undef, sizes[rank+1])

    before_compute = time()
    for t in 1:nb_timesteps
        for i in (1+offsets[rank+1]):(offsets[rank+1] + sizes[rank+1])
            accx = zeros(Float64, nthreads())
            accy = zeros(Float64, nthreads())
            @threads for j in 1:nb_particles
                if i != j
                    dist = distance(particles[i], particles[j]);
                    f = CONST_G*(particles[i].mass * particles[j].mass)/(dist^2);
                    accx[threadid()] += f * (particles[j].posx - particles[i].posx) / dist 
                    accy[threadid()] += f * (particles[j].posy - particles[i].posy) / dist
                end
            end
            my_particles[i - offsets[rank+1]] = particles[i]
            my_particles[i - offsets[rank+1]].accx = sum(accx)
            my_particles[i - offsets[rank+1]].accy = sum(accy)
        end

        @threads for i in 1:sizes[rank+1]
            update_velocity!(my_particles[i])
            update_position!(my_particles[i])
            reset_force!(my_particles[i])
        end

        sendbuf = vcat([Float64[p.mass, p.posx, p.posy, p.velx, p.vely, p.accx, p.accy] for p in my_particles]...)
        recvbuf = MPI.VBuffer(Vector{Float64}(undef, nb_particles*7), sizes .* 7, offsets .* 7)
        MPI.Allgatherv!(sendbuf, recvbuf, comm)
        particles = [Particle(recvbuf.data[7i+1:7i+7]...) for i in 0:nb_particles-1]

    end
    if rank == 0
        fin = time()
        duration_ete = fin - before
        duration_compute = fin - before_compute
        println("$(duration_ete), $(duration_compute), $(particles[1].posx)")
    end
end

main()

File: benchmark_scripts/nbody-mpi/cpp/nbody.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <chrono>
#include <mpi.h>
#include <iomanip>

#define DELTA_TIME 1.0
#define CONST_G 1.0

class Particle {
    public:
        double mass;
        double posx;
        double posy;
        double velx;
        double vely;
        double accx;
        double accy;

        double distance(Particle other);
        void update_position();
        void update_velocity();
        void update_acceleration(Particle other);
        void reset_force();

        Particle(double m, double x, double y, double vx, double vy, double ax, double ay) {
            mass = m;
            posx = x; posy = y;
            velx = vx; vely = vy;
            accx = ax; accy = ay;
        }
};

double Particle::distance(Particle other) {
    return sqrt(pow(posx - other.posx,2) + pow(posy - other.posy,2));
}

void Particle::update_position() {
    posx += velx * DELTA_TIME;
    posy += vely * DELTA_TIME;
}

void Particle::update_velocity() {
    velx += accx/mass * DELTA_TIME;
    vely += accy/mass * DELTA_TIME;
}

void Particle::update_acceleration(Particle other) {
    double dist = this->distance(other);
    double f = CONST_G*(mass * other.mass)/(pow(dist,2));

    accx += f * (other.posx - posx) / dist;
    accy += f * (other.posy - posy) / dist;
}

void Particle::reset_force() {
    accx = 0.0;
    accy = 0.0;
}

double _rand() {
    return (double) std::rand() / RAND_MAX;
}

Particle rand_particle() {
    return Particle(100.0 * _rand(), _rand(), _rand(), _rand(), _rand(), _rand(), _rand());
}

int B[7] = {1, 1, 1, 1, 1, 1, 1};
MPI_Aint D[7] = {
    0,
    1*sizeof(double),
    2*sizeof(double),
    3*sizeof(double),
    4*sizeof(double),
    5*sizeof(double),
    6*sizeof(double)
};
MPI_Datatype T[7] = {
    MPI_DOUBLE,
    MPI_DOUBLE,
    MPI_DOUBLE,
    MPI_DOUBLE,
    MPI_DOUBLE,
    MPI_DOUBLE,
    MPI_DOUBLE,
};

int main(int argc, char** argv) {
    MPI_Init(&argc, &argv);
    // if (argc != 3) {
    //     fprintf(stderr, "USAGE: %s nb_particles nb_timesteps\n", argv[0]);
    //     return 1;
    // }
    int nb_particles = atoi(argv[1]);
    int nb_timesteps = atoi(argv[2]);
    auto begin = std::chrono::high_resolution_clock::now();

    int world;
    int rank;
    MPI_Comm_size(MPI_COMM_WORLD, &world);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    MPI_Datatype mpi_dt_particle;
    MPI_Type_create_struct(7, B, D, T, &mpi_dt_particle);
    MPI_Type_commit(&mpi_dt_particle);

    std::srand(0);
    std::vector<Particle> particles = {};
    particles.reserve(nb_particles);

    if (rank == 0) {
        // is there a more c++ way to do this ?
        for (int i = 0; i < nb_particles; i++) {
            particles.push_back(rand_particle());
        }
    }

    std::vector<int> sizes;
    std::vector<int> offsets;
    int cum_sum = 0;
    for (int k = 0; k < world; k++) {
        sizes.push_back(nb_particles / world);
        offsets.push_back(cum_sum);
        cum_sum += sizes[k];
    }
    sizes[world-1] = nb_particles - (world-1) * (nb_particles/world);
    MPI_Bcast(particles.data(), nb_particles, mpi_dt_particle, 0, MPI_COMM_WORLD);

    std::vector<Particle> my_particles = {};
    my_particles.reserve(sizes[rank]);

    auto begin_compute = std::chrono::high_resolution_clock::now();

    for (int ts = 1; ts <= nb_timesteps; ts++) {
        for (int i = offsets[rank]; i < offsets[rank] + sizes[rank]; i++) {
            double accx = 0.0;
            double accy = 0.0;
#pragma omp parallel for shared(particles,my_particles) reduction(+:accx) reduction(+:accy)
            for (int j = 0; j < nb_particles; j++) {
                if (i != j) {
                    double dist = particles[i].distance(particles[j]);
                    double f = CONST_G*(particles[i].mass * particles[j].mass)/(pow(dist,2));
                    accx += f * (particles[j].posx - particles[i].posx) / dist;
                    accy += f * (particles[j].posy - particles[i].posy) / dist;
                }
            }
            my_particles[i - offsets[rank]] = particles[i];
            my_particles[i - offsets[rank]].accx = accx;
            my_particles[i - offsets[rank]].accy = accy;
        }
#pragma omp parallel for shared(my_particles)
        for (int i = 0; i < sizes[rank]; i++) {
            my_particles[i].update_velocity();
            my_particles[i].update_position();
            my_particles[i].reset_force();
        }

        MPI_Allgatherv(my_particles.data(), sizes[rank], mpi_dt_particle, particles.data(), sizes.data(), offsets.data(), mpi_dt_particle, MPI_COMM_WORLD);
    }
    if (rank == 0) { 
        auto end = std::chrono::high_resolution_clock::now();
        auto duration_compute = std::chrono::duration<float>(end - begin_compute);
        auto duration_ete = std::chrono::duration<float>(end - begin);
        std::cout << duration_ete.count() << "," << duration_compute.count() << "," << std::setprecision (16) << particles[0].posx << std::endl;
    }
    MPI_Finalize();
    return 0;
}
